File: NumberAppearingOnce/NumberAppearingOnce.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        vector<int> res (2,0);
        int number = 0;
        int n = nums.size();
        for (int i=0;i<n;i++)
        {
            number ^= nums[i];
        }
        // lowbit1 : 是二进制表达式中最低位的 1 所对应的值
        int lowbit1 = number&(-number);
        for (int num : nums)
        {
            if ((num & lowbit1) == lowbit1)
            {
                res[0] ^= num;
            }
            else
            {
                res[1]^= num;
            }
        }
        return res;
    }
};

File: NumberAppearingOnce/NumberAppearingOnce.py
class Solution:
    def singleNumbers(self, nums: List[int]) -> List[int]:
        number = 0
        res = [0,0]
        for num in nums:
            number ^= num
        pos = 1
        while (number&pos) == 0:
            pos <<=1
        for num in nums:
            if (num&pos) == pos:
                res[0] ^= num
            else:
                res[1] ^= num
        return res
